File: java/src/io/zihengcat/leetcode/hmap/ContainsDuplicateII.java
package io.zihengcat.leetcode.hmap;

import java.util.Map;
import java.util.HashMap;

/**
 * LeetCode 219. Contains Duplicate II
 * https://leetcode.com/problems/contains-duplicate-ii/
 */
public class ContainsDuplicateII {
    public static void main(String[] args) {
        // ...
    }
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        if (nums == null) {
            return false;
        }
        Map<Integer, Integer> hMap = new HashMap<>();
        Integer nIdx = null;
        for (int i = 0; i < nums.length; i++) {
            nIdx = hMap.get(nums[i]);
            if (nIdx != null) {
                if (i - nIdx.intValue() <= k) {
                    return true;
                }
            }
            hMap.put(nums[i], i);
        }
        return false;
    }
}

/* EOF */

File: cc/hash_table/ContainsDuplicateII.cc
#include <vector>
#include <unordered_map>

using namespace std;

/**
 * LeetCode 219. Contains Duplicate II
 * https://leetcode.com/problems/contains-duplicate-ii/
 */
class ContainsDuplicateII {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        unordered_map<int, int> hMap;
        unordered_map<int, int>::iterator iter;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            iter = hMap.find(nums[i]);
            if (iter != hMap.end()) {
                if (i - iter -> second <= k) {
                    return true;
                }
            }
            hMap[nums[i]] = i;
        }
        return false;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}

/* EOF */
